packages/python/accern_data/accern_data.py

```diff
@@ -273,18 +276,55 @@ def do_init(self, indicator: ProgressIndicator) -> None:
         fname = self.get_path(self._is_by_day)
         indicator.log(f"current file is {fname}")
 
+    def _write_cols(self) -> None:
+        fname = self.get_path(is_by_day=self._is_by_day)
+        if self._cols:
+            with open(f"{fname}.columns", "w") as fout:
+                fout.write(",".join(self._cols))
+
     def add_result(self, signal: pd.DataFrame) -> None:
         fname = self.get_path(is_by_day=self._is_by_day)
+        tmp_fname = f"{fname}.tmp"
         if self._cols is None:
-            signal.to_csv(fname, index=False, header=True, mode="w")
-            self._cols = signal.columns
+            self._cols = signal.columns.to_list()
+            signal.to_csv(tmp_fname, index=False, header=True, mode="w")
         else:
+            if len(signal.columns) <= len(self._cols):
+                for col in set(self._cols).difference(signal.columns):
+                    signal[col] = None
+            else:
+                new_cols = set(signal.columns).difference(self._cols)
+                self._cols += new_cols
             signal[self._cols].to_csv(
-                fname, index=False, header=False, mode="a")
+                tmp_fname, index=False, header=False, mode="a")
 
-    def finish_day(self, indicator: ProgressIndicator) -> None:
+    def finish_day(
+            self,
+            indicator: ProgressIndicator,
+            force_finish: bool = False) -> None:
         # csv files are saved by add_result
-        if self._is_by_day:
+        fname = self.get_path(is_by_day=self._is_by_day)
+        tmp_fname = f"{fname}.tmp"
```

Comment:
```suggestion
        tmp_fname = f"{fname}.~tmp"
```

use a function to generate names